core/jvm/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```diff
@@ -38,230 +37,305 @@ import java.util.concurrent.atomic.AtomicReference
-      t,
+      ex,
       "The throwable provided to onError must not be null"
     )
-    subscriber.onError(t)
+    nextState(Error(ex))
   }
 
-  /** Creates a [[Stream]] from this [[Subscriber]]. */
-  def stream(subscribe: F[Unit]): Stream[F, A] =
-    subscriber.stream(subscribe)
-}
+  /** Called by the upstream reactive-streams system when it has finished sending records. */
+  override def onComplete(): Unit =
+    nextState(Complete(canceled = false))
 
-private[flow] object StreamSubscriber {
+  // Interop API.
 
-  /** Instantiates a new [[StreamSubscriber]] for the given buffer size. */
-  def apply[F[_], A](chunkSize: Int)(implicit F: Async[F]): F[StreamSubscriber[F, A]] = {
-    require(chunkSize > 0, "The buffer size MUST be positive")
-    fsm[F, A](chunkSize).map(fsm => new StreamSubscriber(subscriber = fsm))
-  }
+  /** Creates a downstream [[Stream]] from this [[Subscriber]]. */
+  private[flow] def stream(subscribe: F[Unit]): Stream[F, A] = {
+    // Called when downstream has finished consuming records.
+    val finalize =
+      F.delay(nextState(Complete(canceled = true)))
 
-  /** A finite state machine describing the subscriber. */
-  private[flow] trait FSM[F[_], A] {
+    // Producer for downstream.
+    val dequeue1 =
+      F.async[Option[Chunk[Any]]] { cb =>
+        F.delay {
+          nextState(Dequeue(cb))
 
-    /** Receives a subscription from upstream. */
-    def onSubscribe(s: Subscription): Unit
-
-    /** Receives next record from upstream. */
-    def onNext(a: A): Unit
+          Some(finalize)
+        }
+      }
 
-    /** Receives error from upstream. */
-    def onError(t: Throwable): Unit
+    Stream.bracket(subscribe)(_ => finalize) >>
+      Stream
+        .repeatEval(dequeue1)
+        .unNoneTerminate
+        .unchunks
+        .asInstanceOf[Stream[F, A]] // This cast is safe, since all elements come from onNext(a: A).
+  }
 
-    /** Called when upstream has finished sending records. */
-    def onComplete(): Unit
+  // Finite state machine.
 
-    /** Called when downstream has finished consuming records. */
-    def onFinalize: F[Unit]
+  /** Helper to reduce noise when creating unary functions. */
+  private def run(block: => Unit): () => Unit = () => block
 
-    /** Producer for downstream. */
-    def dequeue1: F[Either[Throwable, Option[Chunk[A]]]]
+  /** Runs a single step of the state machine. */
+  private def step(in: Input): State => (State, () => Unit) =
+    in match {
+      case Subscribe(s) => {
+        case Uninitialized(None) =>
+          Idle(s) -> noop
 
-    /** Downstream [[Stream]]. */
-    final def stream(subscribe: F[Unit])(implicit ev: MonadThrow[F]): Stream[F, A] =
-      Stream.bracket(subscribe)(_ => onFinalize) >>
-        Stream
-          .repeatEval(dequeue1)
-          .rethrow
-          .unNoneTerminate
-          .unchunks
-  }
+        case Uninitialized(Some(cb)) =>
+          WaitingOnUpstream(idx = 0, buffer = null, cb, s) -> run {
```

Comment:
Actually, why not allocate the array here? Then we can remove the `idx == 0` conditional below in `OnNext`.

Reply:
Good point. Since there could be significant lag between requesting elements and when they arrive.